edn/src/matcher.rs

```diff
@@ -0,0 +1,576 @@
+        }
+    }
+
+    fn matches_placeholder(pattern: &'a Value) -> Option<(&'a String)> {
+        match *pattern {
+            Value::PlainSymbol(symbols::PlainSymbol(ref s)) => if s.starts_with('?') { Some(s) } else { None },
+            _ => None
+        }
+    }
+}
+
+/// Pattern matcher for EDN values utilizing specified pattern matching rules.
+/// For example, using this with `DefaultPatternMatchingRules`:
+/// * `[_]` matches an arbitrary one-element vector;
+/// * `[_ _]` matches an arbitrary two-element vector;
+/// * `[?x ?x]` matches `[1 1]` and `[#{} #{}]` but not `[1 2]` or `[[] #{}]`;
+struct Matcher<'a> {
+    placeholders: RefCell<HashMap<&'a String, &'a Value>>
+}
+
+impl<'a> Matcher<'a> {
+    /// Creates a Matcher instance.
+    fn new() -> Matcher<'a> {
+        Matcher {
+            placeholders: RefCell::default()
+        }
+    }
+
+    /// Performs pattern matching between two EDN `Value` instances (`value`
+    /// and `pattern`) utilizing a specified pattern matching ruleset `T`.
+    /// Returns true if matching succeeds.
+    pub fn match_with_rules<T>(value: &'a Value, pattern: &'a Value) -> bool
+    where T: PatternMatchingRules<'a, Value> {
+        let matcher = Matcher::new();
+        matcher.match_internal::<T>(value, pattern)
+    }
+
+    /// Recursively traverses two EDN `Value` instances (`value` and `pattern`)
+    /// performing pattern matching. Note that the internal `placeholders` cache
+    /// may not be empty on invocation.
+    fn match_internal<T>(&self, value: &'a Value, pattern: &'a Value) -> bool
+    where T: PatternMatchingRules<'a, Value> {
+        use Value::*;
+
+        if T::matches_any(pattern) {
+            true
+        } else if let Some(symbol) = T::matches_placeholder(pattern) {
+            let mut cache = self.placeholders.borrow_mut();
+            if !cache.contains_key(symbol) {
+                cache.insert(symbol, value); true
+            } else {
+                value == *cache.get(symbol).unwrap()
+            }
+        } else {
+            match (value, pattern) {
+                (&Vector(ref v), &Vector(ref p)) =>
+                    diff_with(v, p, |a, b| self.match_internal::<T>(a, b)).is_none(),
+                (&List(ref v), &List(ref p)) =>
+                    diff_with(v, p, |a, b| self.match_internal::<T>(a, b)).is_none(),
+                (&Set(ref v), &Set(ref p)) =>
+                    v.len() == p.len() &&
+                    v.iter().all(|a| p.iter().any(|b| self.match_internal::<T>(a, b))) &&
+                    p.iter().all(|b| v.iter().any(|a| self.match_internal::<T>(a, b))),
+                (&Map(ref v), &Map(ref p)) =>
+                    v.len() == p.len() &&
+                    v.iter().all(|a| p.iter().any(|b| self.match_internal::<T>(a.0, b.0) && self.match_internal::<T>(a.1, b.1))) &&
+                    p.iter().all(|b| v.iter().any(|a| self.match_internal::<T>(a.0, b.0) && self.match_internal::<T>(a.1, b.1))),
```

Comment:
Wow, you're right, this is far more subtle than I'd thought.  As I was laying awake in the middle of the night, I realized this might need to test all orderings, or do back-tracking or something.  That is, how would you handle matching `#{?a ?b}` against `#{1 2}`?  That is, how do you know you capture the bindings in the "correct" order?

I'm inclined to look at the transactor tests and see if we could just outlaw patterns in unordered collections, at least for a while.  This seems very difficult to get right; I'll look at your tests now to see how you handle this.